File: 06_black_swaption_pricing/cpp/bachelier_swaption.cpp
#include "bachelier_swaption.hpp"

#include <cmath>

BachelierSwaption::BachelierSwaption(
    double N,
    double K,
    double T,
    double sigma,
    ForwardStartingSwap underlying_swap,
    bool is_payer
)
: swap(underlying_swap)
{

    notional = N;

    strike = K;

    expiry = T;

    volatility = sigma;

    payer = is_payer;
}

double BachelierSwaption::normal_pdf(
    double x
){

    return
        std::exp(-0.5*x*x)
        /
        std::sqrt(2*M_PI);
}

double BachelierSwaption::normal_cdf(
    double x
){

    return
        0.5*erfc(-x/std::sqrt(2));
}

double BachelierSwaption::price(
    YieldCurve& curve
){

    double F =
        swap.forward_swap_rate(curve);

    double K = strike;

    double sigma = volatility;

    double T = expiry;

    double A =
        swap.annuity(curve);

    double stddev =
        sigma * std::sqrt(T);

    double d =
        (F-K)/stddev;

    double value;

    if(payer){

        value =
            notional
            * A
            * (
                (F-K)
                * normal_cdf(d)
                +
                stddev
                * normal_pdf(d)
            );

    }else{

        value =
            notional
            * A
            * (
                (K-F)
                * normal_cdf(-d)
                +
                stddev
                * normal_pdf(d)
            );
    }

    return value;
}

File: 06_black_swaption_pricing/python/bachelier_swaption.py
import numpy as np

from scipy.stats import norm


class BachelierSwaption:
    """
    Normal-model swaption pricing.

    Handles:
    ----------
    - low rates
    - negative rates

    Widely used after 2014.
    """

    def __init__(
        self,
        notional,
        strike,
        expiry,
        underlying_swap,
        volatility,
        payer=True
    ):

        self.notional = notional

        self.strike = strike

        self.expiry = expiry

        self.swap = underlying_swap

        self.volatility = volatility

        self.payer = payer

    def price(self, curve):

        F = self.swap.forward_swap_rate(
            curve
        )

        K = self.strike

        sigma = self.volatility

        T = self.expiry

        annuity = self.swap.annuity(
            curve
        )

        stddev = sigma * np.sqrt(T)

        d = (F - K) / stddev

        if self.payer:

            value = (
                self.notional
                * annuity
                * (
                    (F - K)
                    * norm.cdf(d)
                    + stddev
                    * norm.pdf(d)
                )
            )

        else:

            value = (
                self.notional
                * annuity
                * (
                    (K - F)
                    * norm.cdf(-d)
                    + stddev
                    * norm.pdf(d)
                )
            )

        return value